sdlib/d/sync/sharedlock.d

```diff
@@ -0,0 +1,340 @@
+ * existing holders can unlock.
+ */
+struct SharedLock {
+private:
+	shared Mutex mutex;
+	size_t count;
+	enum size_t Exclusive = 1 << (size_t.sizeof * 8 - 1);
+
+public:
+	/**
+	 * Lock unexclusively.
+	 *
+	 * Can be locked by any number of threads unexclusively. An exclusive
+	 * lock cannot be obtained until all unexclusive locks are released.
+	 */
+	void sharedLock() shared {
+		mutex.lock();
+		scope(exit) mutex.unlock();
+
+		(cast(SharedLock*) &this).sharedLockImpl();
+	}
+
+	void sharedUnlock() shared {
+		mutex.lock();
+		scope(exit) mutex.unlock();
+
+		(cast(SharedLock*) &this).exclusiveUnlockImpl();
+	}
+
+	/**
+	 * Lock exclusively.
+	 *
+	 * Upon return, no other threads can hold this locked.
+	 */
+	void exclusiveLock() shared {
+		mutex.lock();
+
+		(cast(SharedLock*) &this).exclusiveLockImpl();
+	}
+
+	void exclusiveUnlock() shared {
+		assert(mutex.isHeld());
+		scope(exit) mutex.unlock();
+
+		(cast(SharedLock*) &this).unlockWriteImpl();
+	}
+
+private:
+	bool noWriteLock() {
+		return count < Exclusive;
+	}
+
+	void sharedLockImpl() {
+		assert(mutex.isHeld());
+		mutex.waitFor(noWriteLock);
+		++count;
+	}
+
+	bool hasExclusiveLock() {
+		return count == Exclusive;
+	}
+
+	void exclusiveLockImpl() {
+		assert(mutex.isHeld());
+		// Wait for no other exclusive lock.
+		mutex.waitFor(noWriteLock);
+		count += Exclusive;
+		mutex.waitFor(hasExclusiveLock);
+	}
+
+	void exclusiveUnlockImpl() {
+		assert(mutex.isHeld());
+		assert((count & ~Exclusive) > 0);
+		--count;
+	}
+
+	void unlockWriteImpl() {
+		assert(mutex.isHeld());
+		assert(count == Exclusive);
+		count = 0;
+	}
+}
+
+unittest sharedLocks {
+	static runThread(void* delegate() dg) {
+		static struct Delegate {
+			void* ctx;
+			void* function(void*) fun;
+		}
+
+		auto x = *(cast(Delegate*) &dg);
+
+		import core.stdc.pthread;
+		pthread_t tid;
+		auto r = pthread_create(&tid, null, x.fun, x.ctx);
+		assert(r == 0, "Failed to create thread!");
+
+		return tid;
+	}
+
+	shared SharedLock lock;
+	assert(lock.count == 0);
+	lock.sharedLock();
+	assert(lock.count == 1);
+	assert(!lock.mutex.isHeld());
+	lock.sharedLock();
+	assert(lock.count == 2);
+	lock.sharedUnlock();
+	assert(lock.count == 1);
+
+	import d.sync.atomic;
+	shared Mutex mutex;
+	shared Atomic!uint numLocked;
+	bool exitThread;
+	void* run() {
+		lock.sharedLock();
+		scope(exit) lock.sharedUnlock();
+		numLocked.fetchAdd(1);
+		mutex.lock();
+		scope(exit) mutex.unlock();
+		bool shouldExit() {
+			return exitThread;
+		}
+
+		mutex.waitFor(shouldExit);
+		return null;
+	}
+
+	import core.stdc.pthread;
+	pthread_t[50] tids;
+	foreach (i; 0 .. tids.length) {
+		tids[i] = runThread(run);
+	}
+
+	{
+		bool allThreadsReady() {
+			return numLocked.load() == tids.length;
+		}
+
+		mutex.lock();
+		scope(exit) mutex.unlock();
+		mutex.waitFor(allThreadsReady);
+	}
+
+	assert(lock.count == 51);
+	{
+		mutex.lock();
+		scope(exit) mutex.unlock();
+		exitThread = true;
+	}
+
```

Comment:
You might want to wait for all the lock to be released, just like you did wait for all of them to be locked.

Reply:
They should be released before the threads exit, so isn't that what the pthread join does?

Reply:
pthread_join waits for the thread to exit. You should see that the lock are unlocked before the that.

Reply:
OK, done.